### PR TITLE
Add python bindings to cudaobjdetect, cudawarping and cudaarithm

### LINKED ISSUE #2159: CUDA HOG does not expose detect methods to other languages
##### System information (version)
- OpenCV => 4.1.0
- Operating System / Platform => Ubuntu 18.04 on Jetson Nano
- Compiler => :grey_question:

##### Detailed description

The `cv::cuda::HOG` class doesn't expose the `detect` and `detectMultiScale` methods to other languages (Java, Python, etc.) I *think* this is due to these methods missing a `CV_WRAP` in [`cudaobjdetect.hpp`](https://github.com/opencv/opencv_contrib/blob/5b434d3c8bb0ab7a62e824bde02bb398f56f31c3/modules/cudaobjdetect/include/opencv2/cudaobjdetect.hpp#L155-L173), however, I'm not a C++ expert so I wasn't sure about it.  

##### Steps to reproduce
```python
import cv2
gpu_hog = cv2.cuda.HOG_create()
gpu_hog.setSVMDetector(gpu_hog.getDefaultPeopleDetector())
gpu_hog.detectMultiScale() # it doesn't matter that we're missing parameters, this will AttributeError
```

### LINKED ISSUE #2393: OpenCV cuda bindings for python seem receive wrong param types for `cudawarping/src/warp.cpp`
##### System information (version)

- OpenCV => 4.1
- Operating System / Platform => Linux Ubuntu 16.04 64 Bit (Nvidia Xavier)
- CUDA => 10.x 

- OpenCV => OpenCV cuda bindings for python seem receive wrong param types for [cudawarping/src/warp.cpp](https://github.com/opencv/opencv_contrib/blob/4.1.2/modules/cudawarping/src/warp.cpp).

##### Detailed description

The parameter `M` for `warpAffine`, `buildWarpAffineMaps`, `warpPerspective`, `buildWarpPerspectiveMaps` is all defined as cv::Mat? 
However, in python binding, they are expected as `cv.cuda_GpuMat`. (ohterwise `TypeError: Expected Ptr<cv::UMat> for argument '%s'` will be raised).
However, pass `cuda_GpuMat` will trigger error at [Mat M = _M.getMat();](https://github.com/opencv/opencv_contrib/blob/4.1.2/modules/cudawarping/src/warp.cpp#L82) with `xxx/opencv/modules/core/src/matrix_wrap.cpp:118: error: (-213:The function/feature is not implemented) You should explicitly call download method for cuda::GpuMat object in function 'getMat_'`

##### Steps to reproduce

 ```python
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-
#
# Demo usage of cuda.warpAffine from OpenCV 4.x

import cv2 as cv
import numpy as np

mask = np.random.randint(low=100, high=255, size=(1000, 2000)).astype(np.uint8)
mapping = np.array([[1.0, 0.0, 400.0], [0.0, 1.0, 300.0]])

image_container = cv.cuda_GpuMat()
mapping_container = cv.cuda_GpuMat()
dsize = (100, 200)
stream = cv.cuda_Stream()

image_container.upload(mask)
mapping_container.upload(mapping)

crop_mask = cv.cuda.warpAffine(src=image_container,
                               M=mapping_container,
                               dsize=dsize,
                               flags=cv.WARP_INVERSE_MAP,
                               stream=stream).download()

cv.imshow('CropMask', crop_mask)
cv.waitKey()

```

##### Possible solutions

I am not so familiar with `InputArray` as once-for-all param type, so, maybe this can be some easy fix for you guys. Currently, I replace `Mat M = _M.getMat();` to `Mat M;  _M.download(M);` to make python happy (and will break cpp codes..)
Another gossip, the cuda version seems no good than CPU version -.-

### PR DESCRIPTION
**Pull request by cudawarped**
**Main PR**: https://github.com/opencv/opencv/pull/16248

resolves #2159 #2393 #2407 
# This pullrequest changes

Adds overloads to C++ functions to generate the missing python bindings for cudaobjdetect, cudawarping, cudaarithm and cudabgsegm functions.

Python tests have been added for the missing functions and all module specific python tests have also been moved here from the main repo.

<cut/>

```
force_builders=Custom
buildworker:Custom=linux-4
build_image:Custom=ubuntu-cuda:18.04
```

### CODE REVIEW

**Comment by cudawarped** [comment]
Moved module specific python cuda tests from main opencv repo.
Updated SURF headers to generate python bindings and modified `cv::cuda::meanShiftSegmentation()` implementation to generate bindings consistent with the python documentation.

**Comment by alalek** [comment]
Could you please re-check `modules/xfeatures2d/misc/python/test/test_cuda_xfeatures2d.py` test / bindings ?

**TBD on GitHub**

**Comment by cudawarped** [comment]
It must be something to do with the way I am using `*this` to call the `operator()`, only problem is, I am not getting the error on my system.  Did you also see this in python 3.7?

If I make push changes, I am guessing all the tests will pass even if error is still present?

**Comment by cudawarped** [comment]
Apologies I committed the old test case, before I introduced the create method.

I have left both the SURF_CUDA constructors unwrapped
```
    //! the default constructor
    SURF_CUDA();
    //! the full constructor taking all the necessary parameters
    explicit SURF_CUDA(double _hessianThreshold, int _nOctaves=4,
         int _nOctaveLayers=2, bool _extended=false, float _keypointsRatio=0.01f, bool _upright = false);
```
meaning that constructing cuSurf as

`cuSurf = cv.cuda_SURF_CUDA.(hessianThreshold,nOctaves,nOctaveLayers,extended,keypointsRatio,upright)`

will construct and empty object and still throw an exception when calling

`cuSurf.detect(cuMat,cv.cuda_GpuMat())`

However this does force the user to call create() allowing python to catch the following exception 
> OpenCV(4.2.0-dev) Error: The function/feature is not implemented (This algorithm is patented and is excluded in this configuration. Set OPENCV_ENABLE_NONFREE CMake option and rebuild the library) in cv::cuda::SURF_CUDA::create, file \modules\xfeatures2d\src\surf.cuda.cpp, line 56
> s

when not compiled with `OPENCV_ENABLE_NONFREE=ON` and appears to be consistent with the behaviour of methods in the main repo. e.g. 

`bgmog2 = cv.BackgroundSubtractorMOG2()`

 works, however

`bgmog2.getDefaultName()`

throws an exception.

**Comment by cudawarped** [comment]
Overloads have been added for mog and mog2 `getBackgroundImage()` and the documentation updated to reflect the type of the transform matrix required by the cudawarping functions.

**Comment by cudawarped** [comment]
Updated KeyPoints in cuda::ORB::Convert python wrapper to be an output argument.

File: modules/cudaobjdetect/include/opencv2/cudaobjdetect.hpp [hunk 1]
```diff
@@ -162,6 +162,22 @@ class CV_EXPORTS_W HOG : public Algorithm
                         std::vector<Point>& found_locations,
                         std::vector<double>* confidences = NULL) = 0;
 
+    CV_WRAP inline virtual void detect(InputArray img,
```
Comment (alalek) [inline modules/cudaobjdetect/include/opencv2/cudaobjdetect.hpp]: > inline virtual

Please remove "virtual" from these "inline" calls.

File: modules/cudaimgproc/src/mssegmentation.cpp [hunk 2]
```diff
@@ -388,7 +388,20 @@ void cv::cuda::meanShiftSegmentation(InputArray _src, OutputArray _dst, int sp,
...
+        GpuMat dstGpuMat = getOutputMat(_dst, src.size(), src.type(), stream);
+        if (stream)
+            dstGpuMat.upload(dst, stream);
+        else
+            dstGpuMat.upload(dst);
+    }
+    else {
+        _dst.create(src.size(), src.type());
+        Mat dstMat = _dst.getMat();
+        dst.copyTo(dstMat);
```
Comment (alalek) [inline modules/cudaimgproc/src/mssegmentation.cpp]: > 3 lines

Did you try just `dst.copyTo(_dst);` call?

File: modules/cudaobjdetect/include/opencv2/cudaobjdetect.hpp (hunk 1 above)
Comment (cudawarped) [reply modules/cudaobjdetect/include/opencv2/cudaobjdetect.hpp]: Removed, sorry I missed those.

File: modules/cudaimgproc/src/mssegmentation.cpp (hunk 2 above)
Comment (cudawarped) [reply modules/cudaimgproc/src/mssegmentation.cpp]: No, but it works perfectly, I wrongly assumed that this worked the same for `Mat` and `GpuMat`.

**Comment by cudawarped** [comment]
@alalek Made suggested changes.

File: modules/cudaimgproc/src/mssegmentation.cpp (hunk 2 above)
Comment (alalek) [reply modules/cudaimgproc/src/mssegmentation.cpp]: I believe it makes sense to improve `GpuMat` support in `.copyTo()` implementation (in a separate PR).

Comment (cudawarped) [reply modules/cudaimgproc/src/mssegmentation.cpp]: OK I can take a look at that once this has been merged.

**Comment by asmorkalov** [comment]
@cudawarped Please remove useless `inline` option in functions that I mentioned before and rework MOG and MOG2 test to check output objects shape at least.

**Comment by alalek** [comment]
> that I mentioned before

where?

File: modules/cudaarithm/include/opencv2/cudaarithm.hpp [hunk 3]
```diff
@@ -274,6 +274,10 @@ CV_EXPORTS_W void bitwise_xor(InputArray src1, InputArray src2, OutputArray dst,
  */
 CV_EXPORTS void rshift(InputArray src, Scalar_<int> val, OutputArray dst, Stream& stream = Stream::Null());
 
+CV_EXPORTS_W inline void rshift(InputArray src, Scalar val, OutputArray dst, Stream& stream = Stream::Null()) {
```
[the remark this answers is not in the document]
Comment (alalek) [reply modules/cudaarithm/include/opencv2/cudaarithm.hpp]: Not sure if we really want this overload in the list of exported functions.

Lets try `CV_WRAP inline` here (if it works).
if don't work, then left current `CV_EXPORTS_W inline` until bindings generator is improved).

File: modules/cudaobjdetect/include/opencv2/cudaobjdetect.hpp (hunk 1 above)
Comment (cudawarped) [reply modules/cudaobjdetect/include/opencv2/cudaobjdetect.hpp]: Am I right in assuming that removing `inline` will mean these don't appear in the shared lib?

Comment (alalek) [reply modules/cudaobjdetect/include/opencv2/cudaobjdetect.hpp]: Exporting in shared libraries is mainly controlled by CV_EXPORTS`.
In this case - through `class CV_EXPORTS_W HOG` declaration.

`inline` is fine here (`virtual` is not).

File: modules/cudaarithm/include/opencv2/cudaarithm.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (cudawarped) [reply modules/cudaarithm/include/opencv2/cudaarithm.hpp]: CV_WRAP appears to work for functions as well as class methods.

File: modules/cudabgsegm/misc/python/test/test_cudabgsegm.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        cuMat = cv.cuda_GpuMat(npMat)
+        cuMatBg = cv.cuda_GpuMat(cuMat.size(),cuMat.type())
+        cuMatFg = cv.cuda_GpuMat(cuMat.size(),cuMat.type())
+
+        mog = cv.cuda.createBackgroundSubtractorMOG()
+        mog.apply(cuMat, lr, cv.cuda.Stream_Null(), cuMatFg)
+        mog.getBackgroundImage(cv.cuda.Stream_Null(),cuMatBg)
+        mog = cv.cuda.createBackgroundSubtractorMOG()
+        self.assertTrue(np.allclose(cuMatFg.download(),mog.apply(cuMat, lr, cv.cuda.Stream_Null()).download()))
+        self.assertTrue(np.allclose(cuMatBg.download(),mog.getBackgroundImage(cv.cuda.Stream_Null()).download()))
```
[the remark this answers is not in the document]
Comment (cudawarped) [reply modules/cudabgsegm/misc/python/test/test_cudabgsegm.py]: Added tests to check size and type of the returned `GpuMat` to verify that its properties have not been altered by the Python wrapper.

**Comment by asmorkalov** [comment]
:+1:

**Comment by AditTuf** [comment]
@cudawarped can you please till me how do I patch this changes on my Jetson Nano which also uses Ubuntu

**Comment by cudawarped** [comment]
> 
> 
> @cudawarped can you please till me how do I patch this changes on my Jetson Nano which also uses Ubuntu

These changes were merged over a year ago.  Using the latest version 4.5.1 will be sufficient.

**Pull request merged**